server/src/main/java/org/opensearch/index/engine/CompositeIndexWriter.java

```diff
@@ -468,16 +466,31 @@ String getCriteriaForDoc(BytesRef key) {
         DisposableIndexWriter computeIndexWriterIfAbsentForCriteria(
             String criteria,
             CheckedBiFunction<String, CriteriaBasedIndexWriterLookup, DisposableIndexWriter, IOException> indexWriterSupplier,
-            ShardId shardId
+            ShardId shardId,
+            int maxRetryOnLookupMapAcquisitionException
         ) {
             boolean success = false;
             CriteriaBasedIndexWriterLookup current = null;
             try {
-                current = getCurrentMap();
+                int counter = 0;
+                while ((current == null || current.isClosed()) && counter < maxRetryOnLookupMapAcquisitionException) {
+                    // This function acquires a first read lock on a map which does not have any write lock present. Current keeps
+                    // on getting rotated during refresh, so there will be one current on which read lock can be obtained.
+                    // Validate that no write lock is applied on the map and the map is not closed. Idea here is write lock was
+                    // never applied on this map as write lock gets only during closing time. We are doing this instead of acquire,
+                    // because acquire can also apply a read lock in case refresh completed and map is closed.
+                    current = this.current.mapReadLock.tryAcquire();
+                    if (current != null && current.isClosed() == true) {
+                        current.mapReadLock.close();
+                        current = null;
+                    }
```

Comment:
Should we move this logic centrally in `close()`.  Also didn't quite understand why the close operation done as a part of write lock acquisition doesn't handle this logic?